src/base/emu-i386/cpu.c

```diff
@@ -257,8 +257,7 @@ static void fpu_io_write(ioport_t port, Bit8u val)
 {
   switch (port) {
   case 0xf0:
-    /* not sure about this */
-    vm86_fpu_state.swd &= ~0x8000;
```

Comment:
But wasn't this also from bochs?
Untrigger of an IRQ is usually happening
as a result of clearing some status bit.

Reply:
No this wasn't from Bochs. I checked the Bochs source code, and Bochs does the equivalent pic_untrigger(13).
Resetting this bit is wrong, it's not related to IRQ13, rather it dates back to the time when the CPU and FPU were seperate and CPUs needed `fwait` which would wait until this (FPU busy) bit was cleared by the FPU.

Reply:
OK but do you want to say that
just untrigger the irq is good idea?
I mean, only 2 possibilities do exist:
- IRQ is edge-delivered (not to confuse
  with edge-triggered), in which case
  my code was correct and doesn't need
  this patch.
- IRQ is level-delivered (still edge-triggered),
  in which case it gets lowered only together
  with some state bit. Maybe that state bit is
  fully internal and is not exported to fxsave,
  but it should be there is you want to change
  the IRQ to level-delivery.

So in bochs is it level-delivered?

Reply:
it just moves the untrigger from 3 other places in the code to the port handler. So the untrigger happens no matter what since IRQ13 invokes INT75, which does the port i/o, it just happens a bit later with this patch.

Or do you mean that we need something similar to `port60_ready` in `src/base/dev/misc/8042.c`.

There is some documentation here https://xem.github.io/minix86/manual/intel-x86-and-64-manual-vol1/o_7281d5ea06a5b67a-427.html
Bochs does this:
https://github.com/bochs-emu/Bochs/blob/master/bochs/iodev/extfpuirq.cc
but they don't handle IGNNE either.

Reply:
Is there any way to ask fpu if
irq13 is pending, w/o unmasking
IRQs to get it actually delivered?

Reply:
https://xem.github.io/minix86/manual/intel-x86-and-64-manual-vol1/o_7281d5ea06a5b67a-426.html

Here is a better doc.
So basically here's how it works:
- You need to untrigger irq when
  the FPU exception is cleared.
  I suppose you need to add an
  fpu exception-clearing instruction
  into 0x75 handler in bios.s.
- 0xf0 untriggers an IRQ in any
  case. But, if the exception wasn't
  cleared before, it would mask it
  with IGNNE.

So in 0xf0 handler you need to
check if there are pending exceptions
and set IGNNE to 1 if so.
Or if you don't want to implement
IGNNE, then just adding the exception-
clearing instruction into bios.s looks
sufficient in addition to your current
change.

Reply:
I think QEMU may do it best (better than Bochs): mainly
https://gitlab.com/qemu-project/qemu/-/blob/master/target/i386/tcg/sysemu/fpu_helper.c
and also
https://gitlab.com/qemu-project/qemu/-/blob/master/target/i386/tcg/fpu_helper.c

So qemu calls `qemu_irq_lower(ferr_irq);` directly, which `pic_untrigger` also calls but after doing some more stuff, you know much more about that than me :)

IGNNE then masks further exceptions even if SE (bit 7) in the status word is set. That's hard to do without emulation since Linux native and KVM operate with CR0.NE=1.

Both Bochs BIOS and SEABIOS do the exact same thing as dosemu for int75, I believe it's the responsibility of e.g. a DOS C runtime to install an exception handler by hooking int2. That exception handler can then clear SE (which will reset IGNNE).

Reply:
> by hooking int2.

int2 as an NMI?

> That's hard to do without emulation since Linux native and KVM operate with CR0.NE=1.

The doc I was pointing to, says:
```
In the recommended circuit, the deactivation of FERR# is used to 
deactivate IGNNE#. If another circuit is used, the software and circuit together must assure that IGNNE# is 
deactivated no later than the exit from the x87 FPU exception handler.
```
We don't have to emulate "recommended"
circuit. :) Let me not "recommend" to implement
it, and implement the second possibility instead,
which is very simple:
In 0xf0 handler you just check SE and set IGNNE
to its value, then lower IRQ.
In the FPU exception handler you don't raise
IRQ if IGNNE still set.

> Both Bochs BIOS and SEABIOS do the exact same thing as dosemu for int75, I believe it's the
> responsibility of e.g. a DOS C runtime to install an exception handler

But in that case you need to implement
IGNNE. Which looks quite simple, too.
BUT if you don't want to do that, adding
clear instr to bios.s looks like an acceptable
hack to me.

Reply:
I don't think 0xf0 writes should read or write the software FPU state; IGNNE is a hardware flag, not part of any FPU software-accessible state.

Here's an example of int2 (yes, same as NMI) actually handling FPE in Open Watcom's runtime
https://github.com/open-watcom/open-watcom-v2/blob/master/bld/clib/fpu/a/fpe87.asm
Because it calls `fnstenv` first which masks exceptions it will run regardless of IGNNE being properly emulated or not. It does however check for SE
```
test    word ptr ENV_SW[BP],ST_EF_ES
```
to distinguish between FPU exceptions and actual NMIs so playing games with that before calling int2 from int75 is a bad idea.

The exception handler could check for IGNNE, but if it doesn't raise IRQ, it'll go right back to the faulting FPU instruction and fault again I think!

Now what the behaviour should be without the C runtime having installed an INT2 handler? I suppose in real DOS (with CR0.NE=0) IGNNE will be set permanently until some FPU code clears SE. I need to check with a little test program how it behaves, if it causes a hang or not.

Reply:
Ok, so reading a bit more an acceptable hack could be to add an `fnclex` instruction in the int75 handler but after calling int2(which could be hooked).

Still need to test a bit though

Reply:
fnclex works. I tested with a little Open Watcom C DOS16 example, FP exceptions are properly caught with its installed FP exception handler, making the DOS program terminate with "Floating point exception".

But if I then force its int2 vector back to the original one it does strange things without fnclex eventually ending in SIGILL at a strange spot. With fnclex added to int75 just before iret it no longer crashes, though gives an incorrect result, because the "fdiv" turned effectively into a no-op, as with unmasked exceptions no result is stored.

Reply:
here's the example https://gist.github.com/166dff1d398339e822af39319546e42b compiled with `wcl -fpi87 fpex.c`
note I hardcoded the original location for int2, would need to be adjusted of course.

Reply:
> I don't think 0xf0 writes should read or write the software FPU state; IGNNE is a hardware flag, not part
> of any FPU software-accessible state.

Yes, the above is correct.
Does it contradict anything of mine?
But I would note that IGNNE, while
being "not part of any FPU software-accessible state",
is still software-accessible (but not
part of any FPU state, so your sentence
is correct).

> to distinguish between FPU exceptions and actual NMIs so playing games with that before calling int2
> from int75 is a bad idea.

Agree.

> The exception handler could check for IGNNE, but if it doesn't raise IRQ, it'll go right back to the faulting
> FPU instruction and fault again I think!

Perhaps we need to emulate IGNNE
by masking all FPU exceptions? Then
there would be no case of calling an
exception handler and not raising IRQ.
By the way, what if irqs are themselves
masked? Now we try to raise an IRQ
but nothing changes and FPU exception
re-triggers? I guess dosemu can stuck
in an exception flood in that case, even
now w/o any mods. Probably we should
check if the IRQ can be raised, and
otherwise just terminate?

> suppose in real DOS (with CR0.NE=0) IGNNE will be set permanently until some FPU code clears SE.

But this is answered in the doc I pointed.
In some circuitry it will be that way, and
in other circuitry it is set until 0xf0 is written.
We need to emulate the later.

Reply:
> The exception handler could check for IGNNE, but if it doesn't raise IRQ, it'll go right back to the
> faulting FPU instruction and fault again I think!

Oh wait a bit.
I have a vague recollection that we
return to the next instruction in that
case. So what you say is impossible.
Not 100% sure, but quite confident. :)

Reply:
something weird is happening for sure, with my example program on devel (where the `pic_untrigger(13)` is called before `pic_request(13)`). I'll need to trace it to see what's exactly happening there.

Reply:
So have you managed to figure
this out?
If we return to the next instruction
after an FPU fault, rather than to
the current one, then implementing
of IGNNE is going to be really trivial.

Reply:
For fault 0x10 (aka SIGFPE) it's returning to the current instruction unfortunately, so it would need instrlen to increase eip. But I think fnclex is ok after int 2. So I followed your suggestion of moving those before the i/o (even if it makes the BIOS different from Bochs/SEABIOS but I think that's ok)

The reason why in my test it went off into something weird is a bug in the FD kernel (and fdpp too), because the code in `irqstack.asm` isn't re-entrant. I had never considered int 75 calling int 2, but the stack is messed up because of it and eventually overwrites crucial DOS memory. Setting `STACKS=0` in `config.sys` avoids all that and then indeed my test is stuck at the FPU instruction (or rather the one right after, `fwait`, but stuck in any case).

Reply:
> For fault 0x10 (aka SIGFPE) it's returning to the current instruction unfortunately,

Well, that answer is at minimum not
universal. Have you checked that with
the FNxx (no-wait) class of an instructions?

> The reason why in my test it went off into something weird is a bug in the FD kernel (and fdpp too), 

Would you please provide a fix to that? :)

Reply:
https://xem.github.io/minix86/manual/intel-x86-and-64-manual-vol3/o_fe12b1e2a880e0ce-229.html
Yes, what you say is correct, and
here it is said pending exceptions
are ignored by FNxx commands.
I suppose FXnn only generate an IRQ
if CR0.NE is not set, but do not raise
an exception when CR0.NE is set...

Reply:
Then we indeed need to check if an
irq can actually be injected, and
terminate otherwise. Or we'll just
hang in a SIGFPE flood.

Reply:
I'll work on a fix for FreeDOS :)

FNxx commands will never cause exceptions, independent of CR0.NE, that's consistent, and not to worry about.
With the FNCLEX in int75, it's ok.

I think the only case where the approach in this PR could go wrong still is if a DOS FPU exception (hooked int2 or hooked int75) handler relies on IGNNE. It'll get stuck on the first non-FN FPU instruction. But Intel docs don't recommend that, and e.g. OW's handler doesn't rely on it either, as it uses FNSTENV first, which also clears exceptions, and after that it's safe.

It's important to get this right, but at the same time, it's been broken all along (if a DOS program hasn't installed an exception handler) and nobody has noticed. Reminded me of this:
http://www.os2museum.com/wp/floating-point-exceptions-and-dos-extenders/
"Needless to say, this saga underscores just how rarely floating-point exceptions are used. 99.9% of the time, the FPU runs with exceptions masked and the extraordinarily complex legacy floating-point exception delivery mechanism does not matter."

Reply:
> I'll work on a fix for FreeDOS :)

Deal, I'll back-port, let me know.

> FNxx commands will never cause exceptions, independent of CR0.NE

Yes but it can still generate an IRQ
when NE clear. But we don't have to
emulate that, obviously.

> int2 or hooked int75) handler relies on IGNNE. It'll get stuck on the first non-FN FPU instruction.

Maybe the only way to emulate IGNNE
is to skip fpu instruction by hands (adding
instr_len). While sounds simple, I'd
rather not... Agreed that fnclex is sufficient.

So what do you think about adding the
check that the IRQ can actually be injected?

Reply:
oh I can see the FreeDOS ML, where people argue about the FreeDOS distribution. The issues and PRs under https://github.com/FDOS/ seem to pretty much ignore that and move at their own pace (slow but overall professional).

The bug in irqstack.asm is universal though, doesn't matter where you run FreeDOS so might as well have the identical fix in fdpp and fdkernel.

Reply:
> So what do you think about adding the
> check that the IRQ can actually be injected?
can be done, would that just be a check to see if IRQ13 hasn't been untriggered yet (i.e. still triggered in the pic), and if so, exit dosemu?

Reply:
> overall professional).

OK still no one uses dosemu there,
and overall, what future do you see
for that kind of freedos? Go to
virtualbox?
I can only think of the future of a
64bit clang-buildable gdb-debuggable
freedos, not more, not less. :)

Reply:
> can be done, would that just be a check to see if IRQ13 hasn't been untriggered yet (i.e. still
> triggered in the pic), and if so, exit dosemu?

Probably, although currently there
are no funcs to read IRR. Another
way of doing that is to check IMR
bit and isset_IF(), which looks more
casual: IMR is accessible to an IO
port and isset_IF() is at your disposal.

Reply:
> > overall professional).
> 
> OK still no one uses dosemu there, and overall, what future do you see for that kind of freedos? Go to virtualbox? I can only think of the future of a 64bit clang-buildable gdb-debuggable freedos, not more, not less. :)

I do use dosemu2 with the latest FreeDOS kernels. I also have contributed to the kernel, SYS, boot loaders, and SHARE.

Reply:
> I do use dosemu2 with the latest FreeDOS kernels.

Unless I am mistaken, you have your
own DOS kernel too. So your use of a
freedos kernel is probably not too frequent.
Though if you know why you use freedos
kernel and not fdpp, please speak up. :)
If there is anything better done in freedos,
fdpp can do that better yet. So tell me what
is it, and get it in fdpp after a while.

Reply:
> > I do use dosemu2 with the latest FreeDOS kernels.
> 
> Unless I am mistaken, you have your own DOS kernel too.

No, actually, the RxDOS kernel is not nearly usable generally. For dosemu2 the biggest missing piece is the lack of FS redirector support. I rarely use it and haven't progressed on it for years. My debugger(s) and TSRs are the only programs that I currently develop.

> So your use of a freedos kernel is probably not too frequent.

Nearly all developing, testing, and building (including decompress-testing) of DOS software on our server and local Linux desktop box is done using a recent FreeDOS kernel, either in dosemu2 or qemu (or, rarely, Virtual Box 6.x). The DOS desktop box still boots MS-DOS 7.10 by default though, but I rarely use it. (The HP 95LX devices use their own MS-DOS 3.20, and I do not know how viable it is to run another kernel on that.)

> Though if you know why you use freedos kernel and not fdpp, please speak up. :) If there is anything better done in freedos, fdpp can do that better yet. So tell me what is it, and get it in fdpp after a while.

Our kernel is portable among PC-compatible machines =) And it is easier for me to modify and debug, though obviously if you babysat me I could probably learn how to do the same or better for fdpp.

Reply:
There are only a few tricks needed
to know when debugging fdpp.
Namely, you need to add a `local.mak`
file into a build dir and write
`EXTRA_DEBUG = 1` there, after which
you'll get an access to most freedos
variables with gdb.
If some variable is still missing, you
can get an info about it with a double
underscode, eg:
```
(gdb) p break_ena
No symbol "break_ena" in current context.
(gdb) p /x __break_ena
$2 = {sym = {<FarPtrBase<char>> = {ptr = {off = 0x337, seg = 0xfa96}}, 
    obj = std::shared_ptr<class ObjIf> (empty) = {get() = 0x0}, nonnull = 0x0}}
(gdb) p lowmem_base[(0xfa96<<4)+0x337]
$3 = 1 '\001'
```

So you look up the address with 2
underscores and then look up the
value via `lowmem_base`.
IIRC Andrew wanted to add a gdb
macros for this all, but that was on
a very early development stages so
I wasn't yet interested.
Andrew, maybe you still want to add
the gdb instrumentation for fdpp?

Reply:
Hi Bart.
I cherry-picked that patch _after_ the
fnclex patches, as in that case it makes
sense. I also added the check if an FPU
irq can be injected, as we discussed
above.